src/proc/constant_evaluator.rs

```diff
@@ -827,6 +839,53 @@ impl ConstantEvaluator<'_> {
 
         self.expressions.append(expr, span)
     }
+
+    /// Return an iterator over the individual components assembled by a
+    /// `Compose` expression.
+    ///
+    /// Given `ty` and `components` from an `Expression::Compose`, return an
+    /// iterator over the components of the resulting value.
+    ///
+    /// Normally, this would just be an iterator over `components`. However,
+    /// `Compose` expressions can concatenate vectors, in which case the i'th
+    /// value being composed is not generally the i'th element of `components`.
+    /// This function consults `ty` to decide if this concatenation is occuring,
+    /// and returns an iterator that produces the components of the result of
+    /// the `Compose` expression in either case.
+    fn flatten_compose<'c>(
+        &'c self,
+        ty: Handle<Type>,
+        components: &'c [Handle<Expression>],
+    ) -> impl Iterator<Item = Handle<Expression>> + 'c {
+        // Returning `impl Iterator` is a bit tricky. We may or may not want to
+        // flatten the components, but we have to settle on a single concrete
+        // type to return. The below is a single iterator chain that handles
+        // both the flattening and non-flattening cases.
+        let (size, is_vector) = if let TypeInner::Vector { size, .. } = self.types[ty].inner {
+            (size as usize, true)
+        } else {
+            (components.len(), false)
+        };
+
+        components
+            .iter()
+            .flat_map(move |component| {
+                if let (
+                    true,
+                    &Expression::Compose {
+                        ty: _,
+                        components: ref subcomponents,
+                    },
+                ) = (is_vector, &self.expressions[*component])
+                {
+                    subcomponents
```

Comment:
We might have to go one level deeper since you could have a `vec4(vec3(vec2(0, 0), 0), 0)`.

Reply:
Oh, good point!